internal/repository/index.go

```diff
@@ -26,27 +28,72 @@ type Index struct {
 }
 
 type indexEntry struct {
-	packID restic.ID
-	offset uint
-	length uint
+	packID *restic.ID
+	offset uint32
+	length uint32
 }
 
 // NewIndex returns a new index.
 func NewIndex() *Index {
 	return &Index{
-		pack:    make(map[restic.BlobHandle][]indexEntry),
-		created: time.Now(),
+		blob:       make(map[restic.BlobHandle]indexEntry),
+		duplicates: make(map[restic.BlobHandle][]indexEntry),
+		created:    time.Now(),
 	}
 }
 
-func (idx *Index) store(blob restic.PackedBlob) {
+// allBlobs return all blobs to the given blobHandle saved in the index
+// also returns if there exists a blob to the handle or not.
+func (idx *Index) allBlobs(h restic.BlobHandle) []indexEntry {
+	blob, ok := idx.blob[h]
+	if !ok {
+		return nil
+	}
+
+	entries, okDup := idx.duplicates[h]
+	if okDup {
+		all := make([]indexEntry, len(entries)+1)
+		all[0] = blob
+		copy(all[1:], entries)
+		return all
+	}
+
+	return []indexEntry{blob}
+}
+
+// getPack gets the pointer to the given pack ID in the array of saved pack IDs
+// if ID is not present, saves it in that array and returns the pointer to the
+// new entry.
+// This procedere allows to use pack IDs which can be easily garbage collected after.
```

Comment:
I don't think getPack is a good name for this method, since it actually adds the id if not found. Maybe addToPacks?

I'm also a bit worried that we now need a linear scan over the packs. Doesn't this mean backups will take quadratic time?

Reply:
You are half right about quadratic time:

- First, this only applies to packs **within** an index file. The number of those should be bounded as index files will be finalized when full enough.
- Second, when #2773 is merged, only `StorePack` is called from backup which just needs to get the PackID once - and this will be a new pack, so the search then should be removed.
- Basically the same applies when called by `Decode`

Ok, I think it is best to simply remove the search and always add the pack ID. Then your proposed name `addToPacks` also fits much better.

Reply:
is changed now.